lib/async_hooks.js

```diff
@@ -207,6 +208,128 @@ class AsyncResource {
   }
 }
 
+// AsyncContext part
+const idResourcePairing = new Map();
+const contexts = new Set();
+const asyncContextHook = createHook({
+  init(asyncId, type, triggerAsyncId, resource) {
+    // save new resource
+    idResourcePairing.set(asyncId, new WeakReference(resource));
+    const triggerResourceWeak = idResourcePairing.get(triggerAsyncId);
+    contexts.forEach((context) => {
+      context._propagate(resource, triggerResourceWeak && triggerResourceWeak.get(), triggerAsyncId);
+    });
+  },
+  destroy(asyncId) {
+    idResourcePairing.delete(asyncId);
+  },
+  promiseResolve(asyncId) {
+    idResourcePairing.delete(asyncId);
+  }
+});
+
+class AsyncContext {
+
+  constructor() {
+    // WeakMap resource -> store
+    this._resourceStore = new WeakMap();
+    this.count = 0;
+    this._firstCalls = new Map();
+  }
+
+  /**
+   * propagate the context from a parent resource to a child one
+   */
+  _propagate(resource, triggerResource, triggerAsyncId) {
+    let store = this._resourceStore.get(triggerResource || {});
+    if (!store) {
+      const record = this._firstCalls.get(triggerAsyncId);
+      if (record instanceof Map) {
+        store = record;
+        // let's weaken the content of _firstCalls.get(triggerAsyncId)
+        // this is to prevent a potential memory leak
+        this._firstCalls.set(triggerAsyncId, new WeakReference(store));
+      }
+      if (record instanceof WeakReference) {
+        store = record.get();
+      }
+    }
+    if (store) {
+      this._resourceStore.set(resource, store);
```

Comment:
Seems a reference of the current store is propagate here. Shoudn't we do a copy here, or copy on write?
I took a look into the .NET [AsyncLocal](https://docs.microsoft.com/en-us/dotnet/api/system.threading.asynclocal-1?view=netframework-4.7.2) (which serves at least a similar purpose for .NET `await`) and they do a copy on write.
Actually every `ExecutionContext` there has a `Map<AsyncLocal, Value>` which gets referenced in every new `ExecutionContext` and setting a new `AsyncLocal.Value` causes that this map is copied.

see also https://github.com/nodejs/node/pull/21313/files#r195229398 for a NodeJs sample related to this.

Reply:
I am afraid this would lead to degraded performances. wdyt?

Reply:
ups, seems I posted my reply at wrong place => https://github.com/nodejs/node/pull/26540#discussion_r267404245